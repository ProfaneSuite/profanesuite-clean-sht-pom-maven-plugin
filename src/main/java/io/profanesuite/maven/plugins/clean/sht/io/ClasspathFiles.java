package io.profanesuite.maven.plugins.clean.sht.io;

import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.Path;
import java.util.Objects;

public final class ClasspathFiles
{

    private ClasspathFiles(){};

    public static Path resourceFromClasspath (final String classpathResource) {
        try {
            final URL resource = ClasspathFiles.class.getResource(classpathResource);
            return Path.of(Objects.requireNonNull(resource).toURI());
        }catch (URISyntaxException exception) {
            throw new RuntimeException(String.format("Failed to load resource [%s]", classpathResource), exception);
        }
    }
}
