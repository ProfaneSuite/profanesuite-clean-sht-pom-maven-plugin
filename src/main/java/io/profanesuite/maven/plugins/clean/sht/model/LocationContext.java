package io.profanesuite.maven.plugins.clean.sht.model;

import io.profanesuite.maven.plugins.clean.sht.Builder;

import java.nio.file.Path;

public class LocationContext {

    /**
     * The internal location of the pom represented as a Path.
     */
    private final Path internalEffectivePomLocation;

    /**
     * The internal location of the pom represented as a Path.
     */
    private final Path internalProductionPomLocation;


    /**
     * Constructor used by the builder to create the context.
     *
     * @param effectivePomLocation pom location as a path
     * @param productionPomLocation pom location as a path
     */
    private LocationContext(final Path effectivePomLocation,
                            final Path productionPomLocation)
    {
        internalEffectivePomLocation = effectivePomLocation;
        internalProductionPomLocation = productionPomLocation;
    }

    public Path getEffectivePomLocation() {
        return internalEffectivePomLocation;
    }

    public Path getProductionPomLocation() {
        return internalProductionPomLocation;
    }

    public static class LocationContextBuilder implements Builder<LocationContext> {

        /**
         * Effective pom location as a String
         */
        private String internalEffectivePomLocation;
        private Path internalEffectivePomLocationPath;
        private String internalProductionPomLocation;
        private Path internalProductionPomLocationPath;


        /**
         *
         * @param effectivePomLocation Where the production pom will be located.
         * @return The context builder
         */
        public LocationContext.LocationContextBuilder withEffectivePomLocation(final String effectivePomLocation) {
            return withEffectivePomLocation(Path.of(effectivePomLocation));
        }

        public LocationContext.LocationContextBuilder withEffectivePomLocation(final Path effectivePomLocation) {
            internalEffectivePomLocationPath = effectivePomLocation.toAbsolutePath();
            this.internalEffectivePomLocation = internalEffectivePomLocationPath.toString();
            return this;
        }

        /**
         *
         * @param productionPomLocation Where the production pom will be located.
         * @return The context builder
         */
        public LocationContext.LocationContextBuilder withProductionPomLocation(final String productionPomLocation) {
            return withProductionPomLocation(Path.of(productionPomLocation));
        }

        public LocationContext.LocationContextBuilder withProductionPomLocation(final Path productionPomLocation) {
            internalProductionPomLocationPath = productionPomLocation.toAbsolutePath();
            this.internalProductionPomLocation = internalProductionPomLocationPath.toString();
            return this;
        }

        @Override
        public LocationContext build() {
            return new LocationContext(internalEffectivePomLocationPath, internalProductionPomLocationPath);
        }
    }

}
