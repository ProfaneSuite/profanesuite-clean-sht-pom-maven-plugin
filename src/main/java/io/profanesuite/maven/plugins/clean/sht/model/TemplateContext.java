package io.profanesuite.maven.plugins.clean.sht.model;

public class TemplateContext {
}
