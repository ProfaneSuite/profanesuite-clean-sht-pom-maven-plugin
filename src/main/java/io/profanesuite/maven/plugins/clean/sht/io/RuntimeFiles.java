package io.profanesuite.maven.plugins.clean.sht.io;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.OpenOption;
import java.nio.file.Path;

public class RuntimeFiles {

    private RuntimeFiles() {}

    public static String readString(final Path resource, final Charset encoding) {
        try {
            return Files.readString(resource, encoding);
        } catch (IOException exception) {
            throw new RuntimeException("Failed to load content", exception);
        }
    }

    public static Path writeString(final Path resource, final String content, final Charset encoding, final OpenOption... options) {
        try {
            Path parent = createDirectories(resource.getParent());

            return Files.writeString(parent.resolve(resource.getFileName()), content, encoding, options);
        } catch (IOException exception) {
            throw new RuntimeException("Failed to write content", exception);
        }
    }

    public static Path createDirectories(final Path directory) {
        try {
            return Files.createDirectories(directory);
        } catch (IOException exception) {
            throw new RuntimeException("Failed to create directories", exception);
        }
    }
}
