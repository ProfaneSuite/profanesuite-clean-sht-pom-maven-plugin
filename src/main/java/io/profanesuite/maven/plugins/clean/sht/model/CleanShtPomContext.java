package io.profanesuite.maven.plugins.clean.sht.model;

/*
 * MIT License
 *
 * Copyright (c) 2022 ProfaneSuite
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import io.profanesuite.maven.plugins.clean.sht.Builder;
import io.profanesuite.maven.plugins.clean.sht.model.LocationContext.LocationContextBuilder;
import io.profanesuite.maven.plugins.clean.sht.pom.Element;
import org.apache.commons.lang3.ArrayUtils;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.nio.charset.UnsupportedCharsetException;
import java.nio.file.Path;
import java.util.*;
import java.util.stream.Stream;

import static java.util.stream.Stream.of;

/**
 * Context which describes what to use and what to produce.
 */
public class CleanShtPomContext
{
    private LocationContext locationContext;

    private Charset encoding;

    private final Set<Element> inclusions = new HashSet<>();

    private final Set<Element> exclusions = new HashSet<>();

    private CleanShtPomContext() {
    }

    /**
     * Productionised pom location.
     *
     * @return pom location represented as a path
     */
    public Path getProductionPomLocation()
    {
        return locationContext.getProductionPomLocation();
    }

    /**
     * Effective pom location.
     *
     * @return pom location represented as a path
     */
    public Path getEffectivePomLocation()
    {
        return locationContext.getEffectivePomLocation();
    }

    public Charset getEncoding() {
        return encoding;
    }

    private void setEncoding(final Charset encoding) {
        this.encoding = encoding;
    }

    public List<Element> getInclusions() {
        return List.copyOf(inclusions);
    }

    public List<Element> getExclusions() {
        return List.copyOf(exclusions);
    }

    private void setLocationContext(final LocationContext locationContext) {
        this.locationContext = locationContext;
    }

    private void addExclusions(final String[] exclusions) {
        addExclusions(Stream.of(exclusions)
                .map(this::findElement)
                .toArray(Element[]::new));
    }

    private void addExclusions(final Element[] exclusions) {
        this.exclusions.addAll(List.of(exclusions));
    }

    private Element findElement(String elementName) {
        return Stream.of(Element.values())
                .filter(element -> element.getElementName().equalsIgnoreCase(elementName))
                .findAny().orElseThrow(() -> new RuntimeException("Could not find element"));
    }


    private void addInclusions(final String[] inclusions) {
        addInclusions(Stream.of(inclusions)
                .map(this::findElement)
                .toArray(Element[]::new));
    }

    private void addInclusions(final Element[] inclusions) {
        this.inclusions.addAll(List.of(inclusions));
    }

    /**
     * Builder used to create and instance of the context.
     */
    public static class CleanShtPomContextBuilder implements Builder<CleanShtPomContext>
    {

        private CleanShtPomContext instance = new CleanShtPomContext();
        private final LocationContextBuilder locationContextBuilder = new LocationContextBuilder();

        @Override
        public CleanShtPomContext build() {
            final CleanShtPomContext result = instance;
            instance = null;

            final Element[] exclusions = defaultExclusions(result)
                    .orElseGet(() -> calculatedExclusions(result));

            final Element[] inclusions =
            Stream.of(Element.values())
                    .filter(element -> !ArrayUtils.contains(exclusions, element))
                            .toArray(Element[]::new);

            result.setEncoding(Optional.ofNullable(result.getEncoding())
                    .or(() -> Optional.of(toCharset(System.getProperty( "file.encoding" ))))
                    .orElse(StandardCharsets.UTF_8));

            result.addExclusions(exclusions);
            result.addInclusions(inclusions);

            result.setLocationContext(locationContextBuilder.build());
            return result;
        }

        private Element[] calculatedExclusions(CleanShtPomContext result) {
            final Element[] inclusions = result.getInclusions().toArray(Element[]::new);
            final Stream<Element> defaultExclusionMinusUserInclusions = of(Element.DEFAULT_EXCLUDE)
                    .filter(element -> !ArrayUtils.contains(inclusions, element));

            return Stream.concat(defaultExclusionMinusUserInclusions, result.getExclusions().stream())
                    .toArray(Element[]::new);
        }

        private Optional<Element[]> defaultExclusions(CleanShtPomContext context) {
            return Optional.of(context)
                    .filter(c -> c.getExclusions().isEmpty())
                    .filter(c -> c.getInclusions().isEmpty())
                    .map(c -> Element.DEFAULT_EXCLUDE);
        }

        /**
         *
         * @param effectivePomLocation Where the production pom will be located.
         * @return The context builder
         */
        public CleanShtPomContextBuilder withEffectivePomLocation(final String effectivePomLocation) {
            locationContextBuilder.withEffectivePomLocation(effectivePomLocation);
            return this;
        }

        public CleanShtPomContextBuilder withEffectivePomLocation(final Path effectivePomLocation) {
            locationContextBuilder.withEffectivePomLocation(effectivePomLocation);
            return this;
        }

        /**
         *
         * @param productionPomLocation Where the production pom will be located.
         * @return The context builder
         */
        public CleanShtPomContextBuilder withProductionPomLocation(final String productionPomLocation) {
            locationContextBuilder.withProductionPomLocation(productionPomLocation);
            return this;
        }

        public CleanShtPomContextBuilder withProductionPomLocation(final Path productionPomLocation) {
            locationContextBuilder.withProductionPomLocation(productionPomLocation);
            return this;
        }

        public CleanShtPomContextBuilder withExclusions(String... exclusions) {
            instance.addExclusions(exclusions);
            return this;
        }

        public CleanShtPomContextBuilder withInclusions(String... inclusions) {
            instance.addInclusions(inclusions);
            return this;
        }

        public CleanShtPomContextBuilder withEncoding(final String encoding) {
            instance.setEncoding(toCharset(encoding));
            return this;
        }

        private Charset toCharset(final String encoding) {
            return Optional.ofNullable(encoding)
                    .filter(Charset::isSupported)
                    .map(Charset::forName)
                    .orElseThrow(
                            () -> new UnsupportedCharsetException(encoding));
        }

    }
}
