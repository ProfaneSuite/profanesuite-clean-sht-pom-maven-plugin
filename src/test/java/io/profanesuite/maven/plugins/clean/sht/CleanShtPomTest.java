package io.profanesuite.maven.plugins.clean.sht;

/*
 * MIT License
 *
 * Copyright (c) 2022 ProfaneSuite
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import com.google.common.jimfs.Jimfs;
import io.profanesuite.maven.plugins.clean.sht.model.CleanShtPomContext;
import io.profanesuite.maven.plugins.clean.sht.model.CleanShtPomContext.CleanShtPomContextBuilder;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;

import static io.profanesuite.maven.plugins.clean.sht.io.LocalFiles.*;

class CleanShtPomTest {

    private final FileSystem fileSystem = Jimfs.newFileSystem();
    private final Path inMemoryRoot = getPathToRoot(fileSystem);
    private final Path validEffectivePom = copyFile(validEffectivePomFromClasspath(), inMemoryRoot.resolve("example").resolve("pom.xml"));

    private final Path expectedProductionPom = inMemoryRoot.resolve("expected").resolve("pom.xml");

    private final CleanShtPomContextBuilder contextBuilder = new CleanShtPomContextBuilder()
            .withEffectivePomLocation(validEffectivePom)
                .withProductionPomLocation(expectedProductionPom);

    @Test
    void will_generate_production_pom() throws IOException {
        final CleanShtPomContext context = contextBuilder
                .build();

        verifyPomWithContext(context);
    }

    @Test
    void will_remove_properties_from_production_pom() throws IOException {
        final CleanShtPomContext context = contextBuilder
                .withExclusions("PropertieS")
                .build();

        verifyPomWithContext(context);
     }

    @Test
    void will_add_parent_to_production_pom() throws IOException {
        final CleanShtPomContext context = contextBuilder
                .withInclusions("paRent")
                .build();

        verifyPomWithContext(context);
    }


    private void verifyPomWithContext(final CleanShtPomContext context) throws IOException {
        new CleanShtPom(context).run();

        final Path productionPom = context.getProductionPomLocation();
        Approvals.verify(Files.readString(productionPom, StandardCharsets.UTF_8), new Options().forFile().withExtension(".xml"));
    }

}