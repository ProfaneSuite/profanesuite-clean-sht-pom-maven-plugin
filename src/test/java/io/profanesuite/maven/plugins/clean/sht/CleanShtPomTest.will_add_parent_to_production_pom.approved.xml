<?xml version="1.0" encoding="UTF-8"?><!--
MIT License

Copyright (c) 2022 ProfaneSuite

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
--><!-- ====================================================================== --><!--                                                                        --><!-- Generated on 2022-10-07T01:55:58                                       --><!--                                                                        --><!-- ====================================================================== --><!-- ====================================================================== --><!--                                                                        --><!-- Effective POM for project                                              --><!-- 'io.profanesuite.maven.plugins:clean-sht-pom-maven-plugin:maven-plugin:0.0.1-SNAPSHOT' --><!--                                                                        --><!-- ====================================================================== --><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <parent>
        <groupId>io.profanesuite.maven.plugins</groupId>
        <artifactId>profanesuite-maven-plugin</artifactId>
        <version>0.0.1-SNAPSHOT</version>
        <relativePath>../profanesuite-maven-plugin</relativePath>
    </parent>
    <groupId>io.profanesuite.maven.plugins</groupId>
    <artifactId>clean-sht-pom-maven-plugin</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <packaging>maven-plugin</packaging>
    <name>Profanesuite clean sht pom Plugin</name>
    <description>Generates a clean pom for uploading with your release artifact.</description>
    <url>https://www.profanesuite.io/plugins/clean-sht-pom-maven-plugin/</url>
    <inceptionYear>2022</inceptionYear>
    <organization>
        <name>The Profane Suite</name>
        <url>https://www.profanesuite.io/</url>
    </organization>
    <licenses>
        <license>
            <name>MIT License</name>
            <url>http://www.opensource.org/licenses/mit-license.php</url>
        </license>
    </licenses>
    <developers>
        <developer>
            <id>abartish</id>
            <name>Arran Bartish</name>
            <email>admin@profanesuite.io</email>
            <roles>
                <role>Founder</role>
            </roles>
            <timezone>-4</timezone>
            <properties>
                <twitter>@arranbartish</twitter>
            </properties>
        </developer>
    </developers>
    <prerequisites>
        <maven>3.2.5</maven>
    </prerequisites>
    <scm>
        <connection>scm:git:https://github.com/ProfaneSuite/clean-sht-pom-maven-plugin.git/profanesuite-maven-plugin/clean-sht-pom-maven-plugin</connection>
        <developerConnection>scm:git:git@github.com:ProfaneSuite/clean-sht-pom-maven-plugin.git/profanesuite-maven-plugin/clean-sht-pom-maven-plugin</developerConnection>
        <tag/>
        <url>https://github.com/ProfaneSuite/clean-sht-pom-maven-plugin/profanesuite-maven-plugin/clean-sht-pom-maven-plugin</url>
    </scm>
    <properties>
        <assembly.tarLongFileMode>posix</assembly.tarLongFileMode>
        <checkstyle.violation.ignore>RedundantThrows,NewlineAtEndOfFile,ParameterNumber,MethodLength,FileLength</checkstyle.violation.ignore>
        <distMgmtSnapshotsName>Profane Suite Development Snapshot Repository</distMgmtSnapshotsName>
        <distMgmtSnapshotsUrl>file:///home/abartish/src/repo/snapshots</distMgmtSnapshotsUrl>
        <gpg.useagent>true</gpg.useagent>
        <javaVersion>8</javaVersion>
        <maven.compiler.source>11</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
        <maven.plugin.tools.version>3.6.4</maven.plugin.tools.version>
        <mavenVersion>3.2.5</mavenVersion>
        <minimalJavaBuildVersion>1.8</minimalJavaBuildVersion>
        <minimalMavenBuildVersion>3.2.5</minimalMavenBuildVersion>
        <organization.logo>https://profanesuite.github.io/profane-logo.png</organization.logo>
        <project.build.outputTimestamp>2022-07-16T16:14:30Z</project.build.outputTimestamp>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
        <resolverVersion>1.0.0.v20140518</resolverVersion>
        <sisuVersion>0.3.5</sisuVersion>
        <slf4jVersion>1.7.5</slf4jVersion>
        <sourceReleaseAssemblyDescriptor>source-release</sourceReleaseAssemblyDescriptor>
        <surefire.version>3.0.0-M7</surefire.version>
    </properties>
    <dependencies>
        <dependency>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-help-plugin</artifactId>
            <version>3.3.0</version>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.maven</groupId>
            <artifactId>maven-plugin-api</artifactId>
            <version>3.2.5</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.maven</groupId>
            <artifactId>maven-core</artifactId>
            <version>3.2.5</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.maven</groupId>
            <artifactId>maven-model</artifactId>
            <version>3.2.5</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.maven</groupId>
            <artifactId>maven-artifact</artifactId>
            <version>3.2.5</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>1.7.5</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.maven.plugin-tools</groupId>
            <artifactId>maven-plugin-annotations</artifactId>
            <version>3.6.4</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.maven.plugin-testing</groupId>
            <artifactId>maven-plugin-testing-harness</artifactId>
            <version>3.3.0</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.maven</groupId>
            <artifactId>maven-compat</artifactId>
            <version>3.2.5</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.mockito</groupId>
            <artifactId>mockito-core</artifactId>
            <version>2.28.2</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>5.9.1</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-nop</artifactId>
            <version>1.7.5</version>
            <scope>test</scope>
        </dependency>
    </dependencies>
</project>
